sdk/storage/Azure.Storage.Files/src/Models/GetFilesAndDirectories.cs

```diff
@@ -0,0 +1,174 @@
+            left.Equals(right);
+
+        /// <summary>
+        /// Check if two GetFilesAndDirectoriesOptions instances are not equal.
+        /// </summary>
+        /// <param name="left">The first instance to compare.</param>
+        /// <param name="right">The second instance to compare.</param>
+        /// <returns>True if they're not equal, false otherwise.</returns>
+        public static bool operator !=(GetFilesAndDirectoriesOptions left, GetFilesAndDirectoriesOptions right) =>
+            !(left == right);
+
+        /// <summary>
+        /// Check if two GetFilesAndDirectoriesOptions instances are equal.
+        /// </summary>
+        /// <param name="obj">The instance to compare to.</param>
+        /// <returns>True if they're equal, false otherwise.</returns>
+        public bool Equals(GetFilesAndDirectoriesOptions other) =>
+            this.ShareSnapshot == other.ShareSnapshot &&
+            this.Prefix == other.Prefix &&
+            this.PageSizeHint == other.PageSizeHint;
+    }
+
+    internal class GetFilesAndDirectoriesAsyncCollection : StorageAsyncCollection<StorageFileItem>
+    {
+        private readonly DirectoryClient _client;
+        private readonly GetFilesAndDirectoriesOptions? _options;
+
+        public GetFilesAndDirectoriesAsyncCollection(
+            DirectoryClient client,
+            GetFilesAndDirectoriesOptions? options,
+            CancellationToken cancellationToken)
+            : base(cancellationToken)
+        {
+            this._client = client;
+            this._options = options;
+            this.PageSizeHint = options?.PageSizeHint;
+        }
+
+        protected override async Task<Page<StorageFileItem>> GetNextPageAsync(
+            string continuationToken,
+            bool isAsync,
+            CancellationToken cancellationToken = default)
+        {
+            var task = this._client.GetFilesAndDirectoriesAsync(
+                continuationToken,
+                this._options,
+                this.PageSizeHint,
+                isAsync,
+                cancellationToken);
+            var response = isAsync ?
+                await task.ConfigureAwait(false) :
+                task.EnsureCompleted(syncOverAsync: true);
+
+            var items = new List<StorageFileItem>();
+            items.AddRange(response.Value.DirectoryItems.Select(d => new StorageFileItem(true, d.Name)));
+            items.AddRange(response.Value.FileItems.Select(f => new StorageFileItem(false, f.Name, f.Properties?.ContentLength)));
+            return new Page<StorageFileItem>(
+                items.ToArray(),
+                response.Value.NextMarker,
+                response.GetRawResponse());
+        }
+    }
+
+    /// <summary>
+    /// Describes a file or directory returned by
+    /// <see cref="DirectoryClient.GetFilesAndDirectoriesAsync"/>.
+    /// </summary>
+    public class StorageFileItem
```

Comment:
On the other hand, this would be good as equatable,

Reply:
#6808